File: sol/solution/0101-0200/0105/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-export([build_tree/2]).
-spec build_tree(Preorder :: [integer()], Inorder :: [integer()]) -> #tree_node{} | null.
build_tree([], _) -> null;
build_tree(_, []) -> null;
build_tree([Root|PreRest], Inorder) ->
    {LeftIn, [_|RightIn]} = lists:splitwith(fun(X) -> X =/= Root end, Inorder),
    LeftLen = length(LeftIn),
    {LeftPre, RightPre} = lists:split(LeftLen, PreRest),
    #tree_node{
        val = Root,
        left = build_tree(LeftPre, LeftIn),
        right = build_tree(RightPre, RightIn)
    }.

File: sol/solution/0101-0200/0105/Solution.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end
defmodule TreeNode do
  @type t :: %__MODULE__{
          val: integer,
          left: TreeNode.t() | nil,
          right: TreeNode.t() | nil
        }
  defstruct val: 0, left: nil, right: nil
end
defmodule Solution do
  @spec build_tree(preorder :: [integer], inorder :: [integer]) :: TreeNode.t() | nil
  def build_tree([], []), do: nil
  def build_tree([root_val | preorder_rest], inorder) do
    {left_inorder, [_root | right_inorder]} = Enum.split_while(inorder, &(&1 != root_val))
    left_count = length(left_inorder)
    {left_preorder, right_preorder} = Enum.split(preorder_rest, left_count)
    %TreeNode{
      val: root_val,
      left: build_tree(left_preorder, left_inorder),
      right: build_tree(right_preorder, right_inorder)
    }
  end
end